moveit_ros/planning/moveit_cpp/src/planning_component.cpp

```diff
@@ -189,7 +216,127 @@ PlanningComponent::PlanSolution PlanningComponent::plan(const PlanRequestParamet
   //  }
   //}
-  return *last_plan_solution_;
+
+  if (update_last_solution)
+  {
+    last_plan_solution_ = plan_solution;
+  }
+  return *plan_solution;
+}
+
+PlanningComponent::PlanSolution PlanningComponent::plan(const MultiPipelinePlanRequestParameters& parameters,
+                                                        SolutionCallbackFunction solution_selection_callback,
+                                                        StoppingCriterionFunction stopping_criterion_callback)
+{
+  // Create solutions container
+  PlanningComponent::PlanSolutions planning_solutions{ parameters.multi_plan_request_parameters.size() };
+  std::vector<std::thread> planning_threads;
+  planning_threads.reserve(parameters.multi_plan_request_parameters.size());
+
+  // Launch planning threads
+  for (auto const& plan_request_parameter : parameters.multi_plan_request_parameters)
+  {
+    auto planning_thread = std::thread([&]() {
+      try
+      {
+        auto solution = plan(plan_request_parameter, false);
+        planning_solutions.pushBack(solution);
+      }
+      catch (const std::exception& e)
+      {
+        RCLCPP_ERROR_STREAM(LOGGER, "Planning pipeline '" << plan_request_parameter.planning_pipeline.c_str()
+                                                          << "' threw exception '" << e.what() << "'");
+        auto plan_solution = std::make_shared<PlanSolution>();
+        plan_solution->error_code = moveit::core::MoveItErrorCode::FAILURE;
+      }
+    });
+    planning_threads.push_back(std::move(planning_thread));
+  }
+
+  // Launch monitor thread if a stopping criterion is provided as function arguments. Otherwise,
+  // the parallel planning process takes as long as the longest planner takes.
+  if (stopping_criterion_callback != nullptr)
+  {
+    auto monitor_thread = std::thread([&]() {
+      while (planning_solutions.getSolutions().size() < planning_threads.size())
+      {
+        if (stopping_criterion_callback(planning_solutions, parameters))
+        {
+          // Terminate planning pipelines
+          for (auto const& plan_request_parameter : parameters.multi_plan_request_parameters)
+          {
+            auto const& planning_pipeline =
+                moveit_cpp_->getPlanningPipelines().at(plan_request_parameter.planning_pipeline);
+            if (planning_pipeline->isActive())
+            {
+              planning_pipeline->terminate();
+            }
+          }
+        }
+      }
+      using namespace std::chrono_literals;
+      std::this_thread::sleep_for(0.1s);
+    });
+    if (monitor_thread.joinable())
+    {
+      monitor_thread.join();
+    }
+  }
+  // Wait for threads to finish
+  for (auto& planning_thread : planning_threads)
+  {
+    if (planning_thread.joinable())
+    {
+      planning_thread.join();
+    }
+  }
+
+  // Select solution
+  // If no custom solution selection callback is provided, choose the shortest path available
+  if (solution_selection_callback == nullptr)
```

Comment:
This block could itself be a SolutionSelectionCallback which you could set as a default argument.